### DISCUSSION TITLE
ABC + abstractmethod and class that returns itself

### DISCUSSION THREAD

**Discussion by Nathanjp91** [python/typing#1401] [General]
I'm working on an ABC with abstractmethods that returns it's own instance. Mypy will complain on these class methods and I'm wondering if there's a way around this.

minimum reproducible example
```python
class Query(ABC):
    
    def __init__(self, n: int = 0):
        self.n = n
    
    def example(self) -> Query:
        return Query(self.n + 1)
        
    @abstractmethod
    def must_implement(self) -> None:
        pass
```

`example` here will error, `Cannot instantiate abstract class "Query" with abstract attribute "must_implement"`. What I want is to signfiy that the example function should return the object class that inherits this base class and fills in the abstractmethods such that the object can be created. Is this a possible interaction between typing and ABC?

**Comment by AlexWaygood** [comment]
If you use `self.__class__` or `type(self)` instead of hardcoding the constructor, the type checker should be fine with it:

```py
from abc import ABC, abstractmethod

class Query(ABC):
    
    def __init__(self, n: int = 0):
        self.n = n
    
    def example(self) -> Query:
        return self.__class__(self.n + 1)
        
    @abstractmethod
    def must_implement(self) -> None:
        pass
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=653b699eb38f5b32251fc64edc98b5b0

(If the `example()` method is always meant to return an instance of whatever `self.__class__` is, `typing.Self` or `typing_extensions.Self` would also probably be a better return annotation for the method.)

**Reply by Nathanjp91** [reply]
That does work, thanks!

**Reply by Nathanjp91** [reply]
Unrelated to the outcome but I pumped this question and code into chatgpt and it came up with a similar solution, albeit it required correcting as the initial solution didn't work but it managed to eventually arrive at this code block to resolve the issue

```python
from abc import ABC, abstractmethod
from typing import TypeVar

T = TypeVar("T", bound="Query")


class Query(ABC):
    def __init__(self, n: int = 0):
        self.n = n

    def example(self: T) -> T:
        return self.__class__(self.n + 1)

    @abstractmethod
    def must_implement(self) -> None:
        pass

```
